qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,849 @@
+        raise PulseError(
+            f"Invalid ITE expression with length {len(ite_lambda_func)}. "
+            "This should be a list of callables for evaluating if, then, else."
+        ) from ex
+
+    if bool(if_func(*values)):
+        if isinstance(then_func, list):
+            # Nested if clause
+            return _evaluate_ite(then_func, *values)
+        return bool(then_func(*values))
+    if isinstance(else_func, list):
+        # Nested if clause
+        return _evaluate_ite(else_func, *values)
+    return bool(else_func(*values))
+
+
+def _lambdify_ite(ite_expr: List["Expr"], params: List["Symbol"]) -> List[Union[Callable, List]]:
+    """Helper function to lambdify ITE-like symbolic expression.
+
+    Args:
+        ite_expr: Symbolic expression which is a list of "IF", "THEN", "ELSE".
+        params: Symbols to be used in the expression.
+
+    Returns:
+        Lambdified ITE-like expression.
+    """
+    import sympy as sym
+
+    lambda_lists = []
+    for expr in ite_expr:
+        if isinstance(expr, list):
+            lambda_lists.append(_lambdify_ite(expr, params))
+        else:
+            lambda_lists.append(sym.lambdify(params, expr))
+    return lambda_lists
+
+
+class EnvelopeDescriptor:
+    """Descriptor of pulse envelope.
+
+    When the descriptor of the symbolic pulse subclass is called first time,
+    it calls sympy lambdify to create callable and cache the function in the descriptor.
+    This improves performance of repeated evaluation of waveform samples.
+    """
+
+    global_envelopes = {}
+
+    def __get__(self, instance, owner) -> Callable:
+        clsname = owner.__name__
+
+        if clsname not in EnvelopeDescriptor.global_envelopes:
+            gendef = getattr(owner, "_define_envelope")
+            source = gendef()
+            self.add_new_pulse(clsname, owner.PARAM_DEF, source)
+
+        return EnvelopeDescriptor.global_envelopes[clsname]
+
+    @classmethod
+    def add_new_pulse(cls, pulse_name: str, params: List[str], expr: "Expr"):
+        """Add new pulse envelope to descriptor.
+
+        Args:
+            pulse_name: Name of pulse subclass.
+            params: List of parameter names.
+            expr: Symbolic expression of pulse envelope.
+        """
+        import sympy as sym
+
+        param_symbols = [sym.Symbol(p) for p in ["t"] + params]
+        cls.global_envelopes[pulse_name] = sym.lambdify(param_symbols, expr)
+
+
+class ConstraintsDescriptor:
+    """Descriptor of pulse parameter constraints.
+
+    A symbolic pulse subclass may provide multiple expressions to validate parameters,
+    thus it generates a list of callables that takes set of pulse parameters.
+
+    When the descriptor of the symbolic pulse subclass is called first time,
+    it calls sympy lambdify to create callable and cache the function in the descriptor.
+    This improves performance of repeated evaluation of waveform samples.
+    """
+
+    global_constraints = {}
```

Comment:
Can we add a comment stating what the purpose of this is? Is this the dict that stores the constraints?